site/content/blog/argocd-carvel-plugin.md

```diff
@@ -0,0 +1,138 @@
+Expose the plugin by patching the `argocd-cm` ConfigMap's `configManagementPlugins`. The command under `generate` is the command that will be run when using the plugin. 
+
+```yaml
+#! argocd-cm-overlay.yml
+#@ load("@ytt:overlay", "overlay")
+#@overlay/match by=overlay.subset({"kind":"ConfigMap", "metadata":{"name":"argocd-cm"}})
+---
+#@overlay/match missing_ok=True
+data:
+  #! TODO append instead of overwrite this
+  configManagementPlugins: |
+    - name: carvel-ytt
+      generate:                      # Command to generate manifests YAML
+        command: ["sh", "-c"]
+        args: ["ytt -f ."]
+```
+
+### Apply the changes to the cluster
+Since these files need to patch the Argo CD configuration, we can apply the ytt overlay when installing Argo CD directly, or as a separate continuous deployment as an exercise left to the reader.
+
+Now, lets install Argo CD Core. To keep setup simple use the core version that does not include the UI. The manifests are available [here](https://argo-cd.readthedocs.io/en/stable/getting_started/#1-install-argo-cd): `https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/core-install.yaml`. 
+
+Run this command to apply the ytt overlays to the Argo CD core manifests and then apply that to the cluster using kapp (or modify for kubectl, if you prefer). The `-y` continues deployment without confirmation. 
+
+```shell
+$ ytt -f argocd-cm-overlay.yml -f  repo-server-overlay.yml -f https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/core-install.yaml  | kapp deploy --app argo --namespace argocd --file - -y
+```
+
+## Create and template an Application with the ytt plugin
+Now, create an Application resource that watches the git repo directory `config-step-2-template/` containing simple ytt templates and data values. If you want to see updates automatically deployed when you make changes, fork this repo and update the `repoURL` reference below. Add `spec.source.plugin` to tell Argo to use the ytt plugin. 
+
+```yaml
+# simple-app-application.yml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: simple-app
+  namespace: argocd
+spec:
+  project: default
+  # this sync policy will deploy the app without confirmation when changes are detected
+  syncPolicy:
+    automated: {}
+  source:
+    repoURL: https://github.com/vmware-tanzu/carvel-simple-app-on-kubernetes.git
+    targetRevision: develop
+    path: config-step-2-template
+
+    # plugin config
+    plugin:
+      name: carvel-ytt
+  destination:
+    server: https://kubernetes.default.svc
+    namespace: default
+```
+Deploy it to the cluster and see that ytt was used to render the templates. 
+```shell
+$ kapp deploy --namespace argocd --app argo-app --file simple-app-application.yml
+```
+See the status of the deployed Application using the `argocd` cli
+```shell
+$ kubectl config set-context --current --namespace=argocd
+$ argocd login --core
+$ argocd app get simple-app
```

Comment:
After running this command after doing the deploy, I get this message from the Application:

```
CONDITION         MESSAGE                                                       LAST TRANSITION
InvalidSpecError  Application referencing project default which does not exist  2022-02-04 15:21:13 -0600 CST
```

It does eventually sync successfully, but wondering if the project needs to be created in advance?

Reply:
The docs indicate its automatic, but there seems to be a slight delay. 
> If unspecified, an application belongs to the default project, which is created automatically and by default, permits deployments from any source repo, to any cluster, and all resource Kinds.

I added a AppProject to manually create it, which seems to fix the issue 👍